### PR TITLE
M plotfile

### PR DESCRIPTION
**Pull request by ajnonaka**
Added the ability to write M to plotfile (Mx_xface, etc... there are 9 different fields)
Various additions to the WarpX class to support M functionality (accessors, checkpointing, AMR)

### CODE REVIEW

**Comment by ajnonaka** [comment]
# ###############################
####### GENERAL PARAMETERS ######
#################################
max_step = 10
plot_int = 1
amr.n_cell = 64 64 64
amr.max_grid_size = 64
amr.blocking_factor = 64
geometry.coord_sys   = 0
geometry.is_periodic = 1  1  1
geometry.prob_lo     =  -32.e-6 -32.e-6 -32.e-6
geometry.prob_hi     =   32.e-6  32.e-6  32.e-6
geometry.prob_lo     =  -32.e-6 -32.e-6 -512.e-6
geometry.prob_hi     =   32.e-6  32.e-6  512.e-6
amr.max_level = 0

#################################
############ NUMERICS ###########
#################################
warpx.verbose = 0
warpx.use_filter = 0
warpx.cfl = 1.
warpx.do_pml = 0

#algo.em_solver_medium = vacuum # vacuum/macroscopic
#algo.macroscopic_sigma_method = backwardeuler # laxwendroff or backwardeuler
#macroscopic.sigma = 0.0
#macroscopic.epsilon = 8.8541878128e-12
#macroscopic.mu = 1.25663706212e-06

particles.nspecies = 0

#################################
############ FIELDS #############
#################################

my_constants.pi = 3.14159265359
my_constants.L = 141.4213562373095e-6
my_constants.c = 299792458.
my_constants.wavelength = 64.e-6

warpx.E_ext_grid_init_style = parse_E_ext_grid_function
warpx.Ez_external_grid_function(x,y,z) = 0.
warpx.Ex_external_grid_function(x,y,z) = 0.
warpx.Ey_external_grid_function(x,y,z) = "1.e5*exp(-z**2/L**2)*cos(2*pi*z/wavelength)"

warpx.B_ext_grid_init_style = parse_B_ext_grid_function
warpx.Bx_external_grid_function(x,y,z)= "-1.e5*exp(-z**2/L**2)*cos(2*pi*z/wavelength)/c"
warpx.By_external_grid_function(x,y,z)= 0.
warpx.Bz_external_grid_function(x,y,z) = 0.

warpx.M_ext_grid_init_style = constant
warpx.M_external_grid = 1. 2. 3.

#################################
############ DIAGNOSTICS #############
#################################

diagnostics.diags_names = diag
diag.period = 100
diag.diag_type = Full

diag.fields_to_plot = Ex Ey Ez Bx By Bz Mx_xface Mx_yface Mx_zface My_xface My_yface My_zface Mz_xface Mz_yface Mz_zface

File: Source/Diagnostics/ComputeDiagFunctors/CellCenterFunctor.H [hunk 1]
```diff
@@ -38,6 +38,9 @@ private:
     int m_lev; /**< level on which mf_src is defined (used in cylindrical) */
     /**< (for cylindrical) whether to average all modes into 1 comp */
     bool m_convertRZmodes2cartesian;
+
+    // starting component value
```
Comment (RevathiJambunathan) [inline Source/Diagnostics/ComputeDiagFunctors/CellCenterFunctor.H]: Use doxygen comment style for automated documentation
/** starting component */

File: Source/Diagnostics/ComputeDiagFunctors/CellCenterFunctor.H [hunk 2]
```diff
@@ -21,7 +21,7 @@ public:
      */
     CellCenterFunctor(const amrex::MultiFab * const mf_src, const int lev,
                       const amrex::IntVect crse_ratio,
-                      bool convertRZmodes2cartesian=true, int ncomp=1);
+                      bool convertRZmodes2cartesian=true, int ncomp=1, int scomp=0);
```
Comment (RevathiJambunathan) [inline Source/Diagnostics/ComputeDiagFunctors/CellCenterFunctor.H:26]: Please can you add that scomp is an input parameter to the function
See comments above this function
\param[in] scomp starting component of mf_src from which values are averaged/interpolated to mf_dst

File: Source/Diagnostics/ComputeDiagFunctors/CellCenterFunctor.cpp [hunk 3]
```diff
@@ -35,6 +36,6 @@ CellCenterFunctor::operator()(amrex::MultiFab& mf_dst, int dcomp) const
 #else
     // In cartesian geometry, coarsen and interpolate from simulation MultiFab, m_mf_src,
     // to output diagnostic MultiFab, mf_dst.
-    CoarsenIO::Coarsen( mf_dst, *m_mf_src, dcomp, 0, nComp(), 0, m_crse_ratio);
+    CoarsenIO::Coarsen( mf_dst, *m_mf_src, dcomp, m_scomp, nComp(), 0, m_crse_ratio);
```
Comment (RevathiJambunathan) [inline Source/Diagnostics/ComputeDiagFunctors/CellCenterFunctor.cpp:39]: Can you add a return function, sComp() in this class and make m_scomp a 'private' member variable.

File: Source/Diagnostics/Diagnostics.cpp [hunk 4]
```diff
@@ -404,6 +404,24 @@ Diagnostics::InitializeFieldFunctors (int lev)
             m_all_field_functors[lev][comp] = std::make_unique<CellCenterFunctor>(warpx.get_pointer_Bfield_aux(lev, 1), lev, m_crse_ratio);
         } else if ( m_varnames[comp] == "Bz" ){
             m_all_field_functors[lev][comp] = std::make_unique<CellCenterFunctor>(warpx.get_pointer_Bfield_aux(lev, 2), lev, m_crse_ratio);
+        } else if ( m_varnames[comp] == "Mx_xface" ){
+            m_all_field_functors[lev][comp] = std::make_unique<CellCenterFunctor>(warpx.get_pointer_Mfield_aux(lev, 0), lev, m_crse_ratio, true, 1, 0);
+        } else if ( m_varnames[comp] == "Mx_yface" ){
+            m_all_field_functors[lev][comp] = std::make_unique<CellCenterFunctor>(warpx.get_pointer_Mfield_aux(lev, 1), lev, m_crse_ratio, true, 1, 0);
```
Comment (RevathiJambunathan) [inline Source/Diagnostics/Diagnostics.cpp:437]: Why is the bool set to true? the boolena converts RZmodes2cartesian. This is specific to 2D simulations. I think for what we have, wrt, M_multifabs, this boolean should be false.
Can you also add a one line comment using the non-doxygen format
// 
to explain why Mx_xface has scomp=0, and My_yface has comp=1, for new users.

Also, if we need cell-centered Mx, wouldn't we average Mx from all faces to the center?
In other words, the user does not need to know the staggering for M_multifab. Ideally, they just specify "Mx" to be plotted. In this case, we can give Mx_xface, Mx_yface, Mx_zface the same dcomp (destination component)?

File: Source/Initialization/WarpXInitData.cpp [hunk 5]
```diff
@@ -454,6 +454,10 @@ WarpX::InitLevelData (int lev, Real /*time*/)
        }
     }
 
+    if (M_ext_grid_s == "parse_m_ext_grid_function") {
+        Abort("WarpXInitData: M field initializtion parser not implemented yet");
```
Comment (RevathiJambunathan) [inline Source/Initialization/WarpXInitData.cpp]: typo initialization

File: Source/WarpX.H [hunk 6]
```diff
@@ -232,22 +241,26 @@ public:
 
     amrex::MultiFab * get_pointer_Efield_cp  (int lev, int direction) const { return Efield_cp[lev][direction].get(); }
     amrex::MultiFab * get_pointer_Bfield_cp  (int lev, int direction) const { return Bfield_cp[lev][direction].get(); }
+    amrex::MultiFab * get_pointer_Mfield_cp  (int lev, int direction) const { return Mfield_cp[lev][direction].get(); }
```
Comment (RevathiJambunathan) [inline Source/WarpX.H:244]: I know Line 237 was not part of this PR.
Typically, comment goes above the code that is describes. (We can also fix this when we submit a PR from evolveM to material_dev)

File: Source/WarpX.cpp [hunk 7]
```diff
@@ -1002,12 +1006,17 @@ WarpX::AllocLevelMFs (int lev, const BoxArray& ba, const DistributionMapping& dm
         Efield_aux[lev][0].reset( new MultiFab(nba,dm,ncomps,ngE));
         Efield_aux[lev][1].reset( new MultiFab(nba,dm,ncomps,ngE));
         Efield_aux[lev][2].reset( new MultiFab(nba,dm,ncomps,ngE));
+
```
Comment (RevathiJambunathan) [inline Source/WarpX.cpp]: Here the M-multifabs are generated for nodal box-arrays. Do we need them there. May be add a comment to mention that currently M-multifabs are not nodal and for most use-cases, we wont need them to be nodal

**Review by RevathiJambunathan — CHANGES_REQUESTED** [review]
Thank you so much Andy for this PR!
I have reviewed only the portions that you have added and not the M-field initialization or evolution. Since it was not part of this PR. A few minor comments here - mostly related to doxygen style commenting. 
Also, we can talk offline about "Mx_xface" or "Mx" to be listed as fields_to_plot.

**Comment by RevathiJambunathan** [comment]
Also, If you look at the automated checks, it fails for source/style. When you click on Details, it shows that currently the end of line whitespace needs to be removed. 
There are some simple commands are also listed -- which you can copy-paste to remove these end of line whitespaces.

Please can you edit the files listed as suggested and push them as well. Thank you :)

File: Source/Diagnostics/ComputeDiagFunctors/CellCenterFunctor.H (hunk 2 above)
Comment (ajnonaka) [reply Source/Diagnostics/ComputeDiagFunctors/CellCenterFunctor.H:26]: fixed with fd8ec0e0830b1de536788e4d22ed5c9e931355a0

File: Source/Diagnostics/ComputeDiagFunctors/CellCenterFunctor.H (hunk 1 above)
Comment (ajnonaka) [reply Source/Diagnostics/ComputeDiagFunctors/CellCenterFunctor.H]: fixed with 47a07f6c1c53846ce3a6747b329ef2844791ffff

File: Source/Diagnostics/ComputeDiagFunctors/CellCenterFunctor.cpp (hunk 3 above)
Comment (ajnonaka) [reply Source/Diagnostics/ComputeDiagFunctors/CellCenterFunctor.cpp:39]: fixed in 2b12b038fb21862e3fec8139b7bae556efe633fa

File: Source/Initialization/WarpXInitData.cpp (hunk 5 above)
Comment (ajnonaka) [reply Source/Initialization/WarpXInitData.cpp]: fixed in 2b93f8e980890f45d7b8d7c9c6321af89706c4d1

File: Source/Diagnostics/Diagnostics.cpp (hunk 4 above)
Comment (ajnonaka) [reply Source/Diagnostics/Diagnostics.cpp:437]: I added the comment in 8d0726f84430fbe53b74f6945a4a9d5967a2ae63.

Regarding the bool set to true, the default value in the constructor is true so I went with that.  By my understanding this flag doesn't make any difference unless we are cylindrical.

Regarding averaging all 6 face values of Mx to the cell-center, that is also a reasonable thing to do.  I wasn't sure which way we ultimately want (maybe both?), so I went with the easier one to implement :)  We can certainly change this behavior in this or a future PR.

File: Source/WarpX.cpp (hunk 7 above)
Comment (ajnonaka) [reply Source/WarpX.cpp]: I removed these 3 lines in abca1819c760a813e34dace178bbf70d784671f3

File: Source/WarpX.H (hunk 6 above)
Comment (ajnonaka) [reply Source/WarpX.H:244]: fixed it anyway :) in bfccd81b480b1e60f06388684c919d81c88a78d5

**Comment by RevathiJambunathan** [comment]
Thanks for this PR! Merged!!
